File: problems/search_a_2d_matrix/solution.py
#                       |
#                       |
# we can also do this, \|/


# with binary search, we were able to do binary search bcz, if we open the matrix it will be a single sorted list


class Solution:
    def searchMatrix(self, mat: List[List[int]], x: int) -> bool:
        if mat:
            m, n = len(mat) , len(mat[0])
            r, c = 0,  len(mat[0]) - 1             
            
            '''
            
                 *  *  *  * (*)  <--start frm this elemenet checking in both directions
                 *  *  *  *  *
                 *  *  *  *  *
                 *  *  *  *  *
                 *  *  *  *  *
            
            
            '''
            
            while r<m and c>=0:
                if mat[r][c]==x:
                    return True
                if mat[r][c]<x:
                    r+=1
                else:
                    c-=1
        return False

File: problems/search_a_2d_matrix/solution.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();
        
        int l = 0;
        int r = m*n-1;
        
        while (l<r){
            int mid = (l+r)/2;
            int x = mid/n;
            int y = mid%n;
            // cout<<matrix[x][y]<<endl;
            if(matrix[x][y]==target){
                return true;
            }
            else if(matrix[x][y]>target){
                r = mid;
            }else{
                l = mid+1;
            }
        }
        
        return matrix[l/n][l%n]==target;
        
        
        
    }
};